### ISSUE TITLE
concerns about email in Accounts List

### ISSUE THREAD

**Issue by kittock** [fedidcg/FedCM#317]
§ 5.2. Accounts List indicates that email is required. 

I am assuming that the purpose of the email field is to provide a human-readable identifier to help the user disambiguate when they have multiple accounts with the IdP (if there is some other purpose, that would be helpful context).

However, it seems conceivable that not all IdPs will use email addresses, especially in the future as email is becoming less common as a means of communication.

Therefore, I am wondering if email should be generalized with something like "human_readable_identifier" which the IdP can populate with an appropriate string: email, phone number, arbitrary username, etc.

**Comment by npm1** [comment]
That is fair. I wonder if we should have a generic field vs having various fields to populate different kinds of information to be displayed in the browser UI?

**Comment by samuelgoto** [comment]
> However, it seems conceivable that not all IdPs will use email addresses, especially in the future as email is becoming less common as a means of communication.

Yeah, I think you are right.

> I wonder if we should have a generic field vs having various fields to populate different kinds of information to be displayed in the browser UI?

Yeah, another option would be to have a `username`-like attribute.

Another suggestion that I think is worth noting is to allow the RP/IDP to choose what they'd like to share selectively. 

Here is a draft of an API surface that would allow that.

https://github.com/fedidcg/FedCM/issues/242#issuecomment-1146493676

**Comment by hlflanagan** [comment]
This document might be of interest. It includes when/why email addresses might be used (though the username should not be collected). These are the requirements that fed SeamlessAccess: https://groups.niso.org/higherlogic/ws/public/download/21892/NISO_RP-27-2019_RA21_Identity_Discovery_and_Persistence.pdf

**Comment by aaronpk** [comment]
+1 for renaming this field to something other than email. It is entirely possible that a user might not have an email address at an IdP, and they may use other kinds of things as an account identifier. `username` would be fine.

**Comment by anderspitman** [comment]
In addition to allowing generic IDs, would it be possible to include the ID type, to allow RPs that support multiple ID types to know how to handle it without trying to guess from the structure?

**Comment by npm1** [comment]
Note that the fact `email` is the name is not necessarily very visible in the UI itself, other than the disclosure text. But allowing `username` and making `email` optional seems reasonable. Not sure I follow the suggestion to include an ID type. Is this about the RP requesting certain data from the IDP? We do have the proposal for RPs to request different fields from the IDP.

**Comment by anderspitman** [comment]
> Note that the fact `email` is the name is not necessarily very visible in the UI itself, other than the disclosure text. But allowing `username` and making `email` optional seems reasonable. Not sure I follow the suggestion to include an ID type. Is this about the RP requesting certain data from the IDP? We do have the proposal for RPs to request different fields from the IDP.

I think I was misunderstanding the purpose of email in the accounts list (didn't read the original issue closely enough). Sounds like it's intended to be user facing to help them pick an account, but not used for any RP processing. In that case it might not be necessary to disambiguate the type of ID.

**Comment by ekovac** [comment]
Would it be worth enumerating types of unique account identifiers in the wild?
See, eg, how Blizzard BattleNet and Discord's user identifiers work. They aren't usernames, they aren't emails, but they are a user-facing identifier (of the form "name#1234".)

**Comment by gffletch** [comment]
If the goal is to allow users to disambiguate, why not let the user attach whatever they want to the identity so they can identify it in the future? This data could be kept on device or maybe by the IDP. This is kind of what happens today with a password manager where the user has multiple entries for the same site. I get to name the entry with my own string and when the password manager is engaged, it shows me the names I've set and not the underlying details.

**Comment by hlflanagan** [comment]
Discussed on [4 February 2025 CG/WG call](https://github.com/fedidcg/meetings/blob/main/2025/2025-02-04-notes.md)

**Comment by TallTed** [comment]
> Would "handle" be a useful disambiguation from "name" here? "Name" is the fluffy, maybe frequently-changing human-facing thing, "handle" is more persistent/long-lived and typically machine-interpreted identifier. That might be too anchored in a specific time and culture though.

Existing use generally flips these, i.e., `handle` (typically specific to a service or application) is more malleable than `name` (typically grounded in a meatspace identifier, though not always the full or formal version of this identifier).

**Comment by cbiesinger** [comment]
See #435 for a proposal to solve this

**Comment by TallTed** [comment]
> It sounds like you are asking for selective disclosure

No, I'm asking that the user identifier not be forced to be their phone number (`tel`), nor their email address (`email`), nor their "full name" (`name`), nor a "handle"/username (`handle`).

> This issue was primarily concerned with what we show to the user, not what gets shared with the RP

Yes.

It is fine with me if the user is forced to supply/choose *any one or more of the fields listed above* as their identifier(s).

I think the user should always be given the option to *select/choose a different identifier/account* than is already known by (and/or logged in with) the RP and/or browser.

**Comment by cbiesinger** [comment]
> I think the user should always be given the option to select/choose a different identifier/account than is already known by (and/or logged in with) the RP and/or browser.

Do you mean IDP here, instead of RP?

Again, if you want control over what gets shared with the RP, that is not what this issue was filed for.

**Comment by npm1** [comment]
But I do think I would like to push back on the notion that the user somehow picks which attributes are shown, there is simply no affordance for this in the UI and I don't think there should be. This is on the IDP to get it right.

**Comment by TallTed** [comment]
> This is on the IDP to get it right

Indeed. That's where I'm pushing, I think.

Imagine if you will, a testing scenario, where I create multiple accounts on a single provider. My "real" name, my telephone, and my email are all the same on all accounts; only the username differs.

As a human, I need to know which identifiers will be displayed going forward, so that I can set username (which must be displayed fully, or at least made fully visible in mouseover text) appropriately to know the difference between these accounts.

**Comment by npm1** [comment]
Sure, and that is all done when the user is creating the account, in the IDP. That has nothing to do with FedCM. In FedCM, this IDP would pass the usernames in the account endpoint, and they would be shown by the user agent.

**Comment by npm1** [comment]
@TallTed are we ok to close this based on the discussion? Or should we add the suggestions I made above, or something else you would want to see in the spec?

**Comment by TallTed** [comment]
Sometimes the split responsibilities lead to more troubles than they relieve...

Whether FedCM, IdP, RP, or browser is responsible — _as a user,_ I need to see differentiators between my multiple IDs in order to choose the right one, and likewise, when creating those IDs I need to know which field(s) will (or should be) displayed in order to tweak, e.g., the username — maybe instead of `TallTed`, I will need to use the terribly unweildy `TallTed-IdP1-acct1`. 

We can say FedCM bears no responsibility here, but if I'm `TallTed` everywhere, how do I prevent the overlap I don't want — say, if someone beats me to `TallTed` on `IdP1` so I'm `TallTed1` there, but I'm still TallTed on all other IdPs, how do I avoid the conflicts on RealName, email, phone from burning me later?

This may have already been addressed, and I'm just not remembering nor finding it now. I'll satisfy myself with this comment, and trust that folks will hear what I'm saying and act on it however appropriate — raising issues and/or PRs, or just standing pat.

**Comment by npm1** [comment]
Ok sounds good. We can always reopen or create a new issue if something arises
